certora/specs/Timelocks.spec

```diff
@@ -0,0 +1,218 @@
+    // 1. Direct execution via executableAt[data] (if already submitted)
+    // 2. Fresh submission at current time with timelock[selector]
+    // 3. Execution after a pending decreaseTimelock takes effect
+// [BUG] Currently there is a bug on the prover for handling msg.data in the hook that's why decreaseTimelock is filtered
+rule earliestExecutionTimeIncreases(env e, env e_next, bytes data, method f, calldataarg args)
+filtered {
+    f -> f.contract == currentContract && !f.isView && f.selector != sig:decreaseTimelock(bytes4, uint256).selector
+}
+{
+    require e_next.block.timestamp >= e.block.timestamp;
+    require executableAt(data) != 0;
+    require data.length >= 36;
+    bytes4 selector = TimelockManagerHelpers.getSelector(data);
+
+    // Compute earliest execution time before interaction
+    mathint execAtValue = to_mathint(executableAt(data)) == 0 ? max_uint256 : to_mathint(executableAt(data));
+    mathint directSubmitTime = require_uint256(e.block.timestamp + timelock(selector));
+    mathint viaDecreaseTime = minDecreaseLock[selector];
+    mathint earliestBefore = min3(execAtValue,directSubmitTime,viaDecreaseTime);
+
+    // Arbitrary function call (any state change)
+    f(e_next, args);
+
+    // Compute earliest execution time after interaction
+    mathint execAtValueAfter = to_mathint(executableAt(data)) == 0 ? max_uint256 : to_mathint(executableAt(data));
+    mathint directSubmitTimeAfter = require_uint256(e_next.block.timestamp + timelock(selector));
+    mathint viaDecreaseTimeAfter = minDecreaseLock[selector];
+    mathint earliestAfter = min3(execAtValueAfter ,directSubmitTimeAfter,viaDecreaseTimeAfter);
+
+    assert earliestAfter >= earliestBefore;
+}
+
+// Submit correctly sets executableAt based on timelock
+rule submitSetsCorrectExecutableAt(env e, bytes data) {
+    bytes4 selector = TimelockManagerHelpers.getSelector(data);
+
+    require e.msg.sender == curator();
+    require executableAt(data) == 0;
+
+    uint256 expectedDelay = extractExpectedDelay(data);
+
+    uint256 timeBefore = e.block.timestamp;
+    submit(e, data);
+
+    assert executableAt(data) == assert_uint256(timeBefore + expectedDelay), "executableAt must equal timestamp + appropriate timelock";
+}
+
+
+// Function must revert if called before submission
+rule cannotExecuteBeforeMinimumTime(env e,method fb,method f, calldataarg args)
+    filtered {
+        fb -> fb.contract == BeforeMinimumTimeChecker && fb.isFallback,
+        f -> functionTimelocked(f) && f.selector != sig:decreaseTimelock(bytes4, uint256).selector
+    }
+{
+    bytes4 selector = to_bytes4(f.selector);
+    require !abdicated(selector);
+
+    // Check currentTime < min(time1, time2)
+    fb(e, args);
+
+    f@withrevert(e, args);
+    assert lastReverted, "Function must revert before minimum executable time";
+}
+
+// Abdicated functions cannot be called
+rule abdicatedFunctionsMustRevert(env e, method f, calldataarg args)
```

Comment:
This one is already verified in AbdicatedFunctions, although it's nice to factorize the `functionTimelocked` helper